crates/turborepo-lib/src/shim.rs

```diff
@@ -227,114 +226,156 @@ pub struct RepoState {
     pub local_turbo_state: Option<LocalTurboState>,
 }
 
-impl RepoState {
-    /// Infers `RepoState` from current directory.
-    ///
-    /// # Arguments
-    ///
-    /// * `current_dir`: Current working directory
-    ///
-    /// returns: Result<RepoState, Error>
-    pub fn infer(current_dir: &Path) -> Result<Self> {
-        // What we look for first are all directories that contain both a `package.json`
-        // and a `turbo.json`.
-        let potential_turbo_roots = current_dir.ancestors().filter(|path| {
-            fs::metadata(path.join("package.json")).is_ok()
-                && fs::metadata(path.join("turbo.json")).is_ok()
-        });
+#[derive(Debug)]
+struct InferInfo {
+    path: PathBuf,
+    has_package_json: bool,
+    has_turbo_json: bool,
+    package_has_workspaces: bool,
+    workspace_globs: Option<Globs>,
+}
 
-        let mut first_package_json_dir = None;
+impl InferInfo {
+    pub fn has_package_json(info: &'_ &InferInfo) -> bool {
+        info.has_package_json
+    }
+    pub fn has_turbo_json(info: &'_ &InferInfo) -> bool {
+        info.has_turbo_json
+    }
 
-        // We loop through these directories and see if there are workspaces defined in
-        // them, either in the `package.json` or `pnm-workspaces.yml`
-        for dir in potential_turbo_roots {
-            if first_package_json_dir.is_none() {
-                first_package_json_dir = Some(dir)
-            }
+    pub fn is_workspace_root_of(&self, target_path: PathBuf) -> bool {
+        match &self.workspace_globs {
+            Some(globs) => globs
+                .test(self.path.to_path_buf(), target_path)
+                .unwrap_or(false),
+            None => false,
+        }
+    }
+}
 
-            let pnpm = PackageManager::Pnpm;
-            let npm = PackageManager::Npm;
-            let is_workspace =
-                pnpm.get_workspace_globs(dir).is_ok() || npm.get_workspace_globs(dir).is_ok();
+impl RepoState {
+    fn generate_potential_turbo_roots(reference_dir: &Path) -> Vec<InferInfo> {
+        // Find all directories that contain a `package.json` or a `turbo.json`.
+        // Gather a bit of additional metadata about them.
+        let potential_turbo_roots: Vec<_> = reference_dir
+            .ancestors()
+            .map(|path| {
+                let workspace_globs = PackageManager::Pnpm
+                    .get_workspace_globs(path)
+                    .unwrap_or_else(|_| {
+                        PackageManager::Npm
+                            .get_workspace_globs(path)
+                            .unwrap_or(None)
+                    });
+
+                InferInfo {
```

Comment:
If I'm reading this correctly, we allocate a `PathBuf` on every ancestor, but then filter on `has_package_json` and `has_turbo_json`. Could we use `filter_map` and avoid these extra allocations by early returning `None` if `has_package_json` and `has_turbo_json` are both false?

Reply:
This is a function I didn't know existed!